### ISSUE TITLE
Transition a Master to an available Shadow

### ISSUE THREAD

**Issue by cratoo** [lizardfs/lizardfs#613]
I built a HA-solution based on etcd (with test-and-set), which basically works great and as expected.

I'm struggling right now with, that there does not seem to be an option to convert a running master to a available shadow.

I tried to use the personality 'ha-cluster-managed', which allows me to do:
`lizardfs-admin promote-shadow localhost 9421`

But if I want to "demote" a master, I only see the options:
* Changing the config to shadow and start
* running `mfsmaster -o ha-cluster-managed -o initial-personality=shadow -c /path/to/config`

In both cases the shadow comes up, but has a metadataversion of 0 and therefore is not usable as it never had a connection to a master before.

Isn't there an option to make a master to a available shadow or did I miss something?

**Comment by Zorlin** [comment]
Not without a restart.

**Comment by guestisp** [comment]
@cratoo interesting solution with etcd, I was thinking the same some days ago.

Is your configuration private or can be published somewhere?

**Comment by cratoo** [comment]
@guestisp currently my productive LFS clusters still even run with a bash-script for a failover. But this turned out to be "interesting".
As soon as I consider the python-Script, which takes care for lfs-migration based on the etcd test-and-set, as mature I can share it.

Basically it works like this on each lfs master/shadow node:
In a loop:
* check etcd cluster health
* if etcd broken
  * make myself to shadow
* if etcd ok
  * if lfs is master or shadow:
    * master -> update etcd master key with a 10s ttl to the master hostname
    * shadow -> test and set etcd master key none to my hostname as master hostname
      * if it works we're new master and a failover had to happen.

Obviously at some point I also need to add/remove a floating ip, which I didnt mention here to keep the logic as simple as possible.

**Comment by guestisp** [comment]
Why are you setting master as the current node? What if you have multiple shadows? All shadow will became master as all shadow will see a failed etcd

**Comment by cratoo** [comment]
etcd guarantees for test-and-set that only one call will succeed.
And yes I have multiple shadows, actually it is mostly a 4x master/shadow setup.

**Comment by guestisp** [comment]
Really interesting. Why not publish the Python script even if unfinished? Maybe someone could help with bugfix

**Comment by cratoo** [comment]
As said, it already works great. And yesterday I already even built a workaround for this transition issue.
I didn't publish it yet, because there are still dependencies in it that depend on our infrastructure/tooling.
What I can do is publish the script next week and basically replace our dependencies with '# call something that does something', if that helps.

**Comment by guestisp** [comment]
That would be great as starting point for HA in lizard.
Etcd is very easy to setup and configure and automatic failover with multiple servers without using the over bloated corosync would be awesome

**Comment by jkiebzak** [comment]
I'm definitely interested too.

**Comment by guestisp** [comment]
With consoul should be even easier, as there is a native locking system that allows one or more servers to hold a resource with a global lock.

After leader election, consoul automatically start an app on the leader server and automatically kill that when another leader is elected.

Next week I'll make some tests

**Comment by 4Dolio** [comment]
This https://github.com/lizardfs/lizardfs/issues/598#issuecomment-334202503 may be of interest.

There is no promote master operation.  You have to use initial-personality=shadow.  At that point any of that masters metadata is (almost) useless, so it ignores any metadata it might have and begins from 0.  It asks the master for a full copy of metadata and begins metalogging from that point.  It loads the full set into memory and then replays the changelog to catchup to $now.  At that point the shadow is ready to be promoted as needed.  This process should be rather fast unless your metadata set if massive.  I have one such metadata set which can take anywhere from 15 to 30 minutes to become completely "ready" and caught up to the current master.  Most of my smaller metadata sets are ready in under 30 seconds once started with initial-personality=shadow

**Comment by 4Dolio** [comment]
Another quark.. When killing a the master:master there is a special call to request it "quick stop" without writing it's metadata to disk (which could be slow).  I have found that a promoted master:shadow will not take over until the previous master is no longer connected, which would be delayed if the master were to write to disk before exiting. .oO( The master could/might be changed to drop shadow connections and then write it's data, allowing for the shadow to take over more quickly, perhaps).  But that point is mostly moot because only the current master metadata stream is valid, so all other copies can/should be discarded anyway.  Of course this means great care is taken to ensure no split-brain (this the shadow waiting on full exit of old master).  When a master is quick-stopped, it will know that it's metadata is unclean and require a restore to make it valid again (if you needed it for some special reason).

When a shadow stops, it also writes it's metadata to disk, This metadata is valid and could be used to seed a initial-personality=master within the cluster with no repair intervention.  As such my HA process also renamed a shadow metadata such that it is not silently valid and will not accidentally seed the cluster.  If shadows also had a quick-stop function I would use that instead of renaming the metadata by hand upon the completion of a stop of a shadow.

**Comment by 4Dolio** [comment]
One last note about my (https://github.com/4Dolio/lizardfs/tree/4Dolio-corosync-ocf-metadataserver-patch-1/src/ha-cluster) ocf metadataserver script.  It errors on the side of not transitioning during some faults.  The most recent revisions may have negated some fault detections?  In the end, I prefer this bias in corosync because I have found that corosync itself may sometimes fault, but not transition (Due to recent changes to ignore some faults).  It turns out I prefer this behaviour because the master turns out to be completely healthy in those cases and I am glad that it did not transition needlessly.  However this sometimes leaves a shadow (resumed to working perfectly but) in an error state according to corosync.  So corosync will not promote a shadow in such a state, I use corosyncs tools to clear errors in such cases.  

I have also noticed that corosync sometimes fails to be able to run the metadataserver checks because the corosync master node (not to be confused with the LFS master/shadow) had become jacked up somehow and is no longer controlling the cluster, and no other nodes take over in this case.  To solve this I kill all corosync and pacemaker processes, which safely leaves the LizardFS master alone and operational.  This lets some other corosync node take over control.  Then I resume the faulty corosync services, which again leaves the running LizardFS processes alone.  Corosync does a good job of detecting the LFS master state if it happens to already be running.  This again normally needs to clear the errors that resulted from the killing of the zombie corosync.

These exceptions are fairly rare and easy to deal with once you know how to handle them.  I should enhance the documentation of my metadataserver script perhaps...

**Comment by guestisp** [comment]
With consul I was able to create a relatively simple multi-node failover.

lizardfs-master set to run a **shadow** on every node.

i've defined a consul watcher for key `leader` that runs a transition script on every change of value.
The transition script checks if value of key `leader` equals current hostname. If it's the current hostname, will promote local lizardfs to master and add the floating-ip, else, if it was a master, will force a demotion (with a restart) and remove the floating-ip

Next, with Consul global lock, i'm running a very, very simple bash script on each node that will set the local hostname as current leader in key `leader`, triggering the watcher. Consul run this script only once and only on a single node, thus, only one node will be promoted.

I have to understand how to automatically remove a node from consul if a service is failed (in example, lizardfs-master crashed, in this case, the node should not compete to leader election)

**Comment by Blackpaw** [comment]
>The transition script checks if value of key leader equals current hostname. If it's the current hostname, will promote local lizardfs to master and add the floating-ip, else, if it was a master, will force a demotion (with a restart) and remove the floating-ip

How are you managing the floating ip?

**Comment by guestisp** [comment]
I don't use any floating ip anymore
I'm using DNS server provided by Consul that is always up to date with the master IP

I'm also trying to see what perform better between changing the hosts file or changing the configuration

**Comment by guestisp** [comment]
Other than this, you can simply use "ip" command to add and remove the IP from metadata nodes

**Comment by Blackpaw** [comment]
>you can simply use "ip" command to add and remove the IP from metadata nodes

Ah, of course, thanks.

>I'm using DNS server provided by Consul that is always up to date with the master IP

Do the lizard service such as chunkservers and mfsmount do a dns lookup when reconnecting or do they just use the old ip?

>I'm also trying to see what perform better between changing the hosts file

Is that rather than using the consul DNS server?

**Comment by guestisp** [comment]
> Ah, of course, thanks.
So, what's the issue with that? A floating IP is simple an IP moved around. Just add the IP on a new machine and, just to be sure, issue some gratuitous arp `arping -c 10 -A -I eth0 10.0.0.1`

> Do the lizard service such as chunkservers and mfsmount do a dns lookup when reconnecting or do they just use the old ip?

I'm trying to figure this out, but i'm stuck here (the same happens even by changing the IP in mfsmaster.cnf and reloading master):
https://github.com/lizardfs/lizardfs/issues/623#issuecomment-344633047
Keep in mind that Consul DNS replies are already expired (TTL set to 0), thus no cache is involved client side.


> Is that rather than using the consul DNS server?

Yes. It was just a test to see if changing the mfsmaster.cnf would fix the issue reported above, but it doesn't seem so....

**Comment by 4Dolio** [comment]
That sound very similar to how MooseFS handles failover.   Though i have not used modern MooseFS and have not used this method I think I prefer a floating IP.

The reason being that DNS in an of itself is yet another service dependency.  There are numerous other name caching mechanics to consider when using this method.  All participants (CS, Clients, etc) must have perfect DNS resolution, while I prefer the storage work even without and DNS.

Anyway, I prefer to avoid all this added complexity and stick to IP for all members including CS and Clients.

**Comment by guestisp** [comment]
You are right, but our orchestrator manages DNS it self, thus there is no an external dependancy.

But i'm trying all available methods (dns, configuration change, hosts file replacement, floating IP, ....) to just see which one performs better.

**Comment by Blackpaw** [comment]
Thinking on it, I doubt dynamic dns lookups, hosts file changes etc would work with lizardfs for changing the mfsmaster ip on the fly - most apps for performance reasons would be caching their dns resolutions. But certainly worth testing.

Would be helpful if skytech could pass an opinion on this.

**Comment by 4Dolio** [comment]
Always good to explore all your options.. just wanted to point out potential additional complexity and problems..

**Comment by Blackpaw** [comment]
True, I'm intrigued by consul (new one to me, thanks @guestisp) as it seems very reliable, predictable and well documented. Relatively simple to get going, but powerful and flexible. Getting it to drive lizard master servers is the interesting bit.

keepalived is ok, but for me proving difficult on the edge cases and its documentation is shocking. The whole project is a little to adhoc for my taste.

**Comment by guestisp** [comment]
@Blackpaw I've almost done with consul. Do you have a test cluster? I can provide you our POC in next days if you like

**Comment by Blackpaw** [comment]
That would be great thanks. Have a virtual test cluster running.

**Comment by guestisp** [comment]
> That would be great thanks. Have a virtual test cluster running.

Ok, when I've figured out why shadow can't be promoted, i'll send you a POC in (very) alpha stage.

**Comment by hoonetorg** [comment]
@guestisp how is the status of consul/lizardfs? i would like to help you.

**Comment by guestisp** [comment]
Temporary in hold due to Christmas vacation and some other urgent needs. I think to continue working with Consul in January or February

**Comment by hoonetorg** [comment]
hi @guestisp, sent you my contact information over the contact form at https://www.guest.it/contatti (hope it reaches you).
Contact me if you want and when you are back in business :).

**Comment by baby-gnu** [comment]
Hello.

This is an interesting thread, does someone have public repository about HA setup with consul/etcd?

Regards.

**Comment by baby-gnu** [comment]
It seems that every body abandoned using etcd/consul with lizardfs, right @guestisp and @cratoo ?

Regards.

**Comment by guestisp** [comment]
I've abandoned lizardfs at all

**Comment by baby-gnu** [comment]
Ok @guestisp, sorry for the noise.
